drivers: led_strip: add WS2812 I2S-based driver

narvalotech opened the pull request:
Add a driver implementation that uses the I2S peripheral. Based off this blog post:
https://electronut.in/nrf52-i2s-ws2812/

Should help with #33505, #29877 and maybe #47780, as there is no garbage data at the end of transmissions on nRF52832, and no gaps.

narvalotech committed:
drivers: led_strip: add WS2812 I2S-based driver

Add a driver implementation that uses the I2S peripheral.
Based off this blog post:
https://electronut.in/nrf52-i2s-ws2812/

Should help with #33505, #29877 and maybe #47780, as there is no garbage
data at the end of transmissions on nRF52832, and no gaps.

Signed-off-by: Jonathan Rico <jonathan@rico.live>

narvalotech commented:
[logic-traces.zip](https://github.com/zephyrproject-rtos/zephyr/files/9955367/logic-traces.zip)
attaching some traces taken with the two different drivers on a thingy52 (nRF52832), they can be opened with [DSView](https://github.com/DreamSourceLab/DSView).

simonguinot commented:
Hi @narvalotech. Thanks for this alternative driver. I'll review it this week-end.

mbolivar-nordic commented on dts/bindings/led_strip/worldsemi,ws2812-i2s.yaml:
```diff
@@ -0,0 +1,24 @@
+
+  out-active-low:
+    type: boolean
+    required: false
```
Remove this. False is the default and true would not make sense for booleans.

narvalotech replied:
yeah boolean handling in dts is still a bit weird to me

mbolivar-nordic commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,255 @@
+     * Convert pixel data into I2S frames. Each frame has pixel data
+     * in color mapping on-wire format (e.g. GRB, GRBW, RGB, etc).
+     */
+    for (uint16_t i = 0; i < num_pixels; i++) {
+        for (uint16_t j = 0; j < cfg->num_colors; j++) {
```
Why are you using uint16_t in some places and int in others?  This inconsistency is weird.

narvalotech replied:
agreed, I'll fix it

narvalotech replied:
oh yeah I remember now, it's because comparison between signed and unsigned is a bit hairy. 
I'll change the only `int` iterator to use `uint16_t` if that's ok.

mbolivar-nordic commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,255 @@
+
+#define WS2812_I2S_DEVICE(idx)                                          \
+                                                                        \
+    K_MEM_SLAB_DEFINE_STATIC(ws2812_i2s_##idx##_slab,                   \
```
Why bother with this instead of just allocating an array per device?

narvalotech replied:
I think it's a requirement of the I2S API (which is not super-flexible)
https://github.com/jori-nordic/zephyr/blob/969d3ffe92e806ffb32fa6d81ecca3cd1f398e09/include/zephyr/drivers/i2s.h#L279

narvalotech replied:
another issue with that API is that I can't really check if a transaction's status, hence the sleep duration in the led update implementation.
I could make a retry loop based on the error message of `i2s_write` but I think waiting is cleaner.

If you can stress-test it a bit it would be nice, ie just call the led update function in a loop from your app and see what happens.

mbolivar-nordic reviewed (changes requested):
I'm away from my devices and will test next week, but I'm glad to see this PR, thanks.

Please fix the compliance failures.

A couple of initial things I noticed.

narvalotech commented:
Thanks for reviewing, I'll make an update after you two have tested it, in case I need to fix any other issues.

simonguinot commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,255 @@
+
+static const struct led_strip_driver_api ws2812_i2s_api = {
+.update_rgb = ws2812_strip_update_rgb,
+.update_channels = ws2812_strip_update_channels,
```
Some tabs are missing above.

Please run checkpatch.pl against the driver and fixes the whole indentation. Tabulations are expanded into 4 spaces everywhere...

narvalotech replied:
yep sorry about that, I haven't yet installed the editorconfig plugin on my personal machine, I'll fix the compliance issues.

simonguinot commented on samples/drivers/led_ws2812/boards/thingy52_nrf52832.conf:
```diff
@@ -0,0 +1,5 @@
+
+CONFIG_I2S=y
+CONFIG_WS2812_STRIP=y
+CONFIG_WS2812_STRIP_I2S=y
```
I think the sample is getting a bit messier.  And it already was some :)

We have configuration and overlay files using different drivers. But it is impossible to tell which one from the filenames...

@mbolivar-nordic, probably there is something to do here ?

narvalotech replied:
should we add a section in the readme for each? at least one for each driver.

simonguinot replied:
I am not sure how we should deal with that. @mbolivar-nordic, what do you think ?

narvalotech replied:
added.

simonguinot reviewed (changes requested):
Hi @narvalotech,

Again thanks for this driver and sorry for the delay.

Note that I didn't review it fully but I already have some comments / questions for you.

The most important is that I'd like to see this driver configurable (like `ws2812_spi`) to support more controller models (such as the Everlight B1414).

In the meantime I'll manage to get a board with an I2S port. And I'll try to test it with a B1414...

narvalotech commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,255 @@
+        sym_one = 0xE;
+        sym_zero = 0x8;
+        reset_word = 0;
+    }
```
ah so that's what it's for! I thought it was just to correct for the SPI timings that might differ a bit from manufacturer to manufacturer. 
Yeah I'll add `nibble-one` and `nibble-zero` properties (since it's 4-bits per symbol). I'll still keep the `active-low` prop though because that is useful for governing the reset and idle logic values.

simonguinot replied:
That's also correct. The SPI frequencies available can be different from a controller to another. So it serves both purposes.

simonguinot replied:
OK. With the current static configuration the timings are almost perfect for the Everlight B1414 controller:
```
0 bit: 300 ns high and 900 ns low.
1 bit: 900 ns high and 300 ns low.
```
While I still think it is interesting to allow the user to configure the frames of the 0 and 1 bits and the clock frequency (because there are probably almost compatible devices with other specifications out there), I also think it can be added later. So I won't block this PR if you don't add it now.

narvalotech commented:
here's what prompted this btw https://jonathan.rico.live/projects/jenkins-juletre/

simonguinot commented:
Hi @narvalotech,

I finally managed to test your driver with an NXP LPC55S69 board and a strip with 18 Everlight B1414 LEDs. And except an issue with the reset/latch delay, I confirm it is working. Nice work :)

simonguinot commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,246 @@
+#include <zephyr/sys/util.h>
+
+#define WS2812_I2S_PRE_DELAY_WORDS 1
+#define WS2812_I2S_LRCK_HZ	   100000
```
I think it would be helpful to have a comment somewhere explaining the relation between the WS clock, the MS clock, and the four bits describing the encoding of the zero and one frames.

simonguinot commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,246 @@
+#include <zephyr/sys/util.h>
+
+#define WS2812_I2S_PRE_DELAY_WORDS 1
+#define WS2812_I2S_LRCK_HZ	   100000
```
Also, it would be nice to allow the user to configure the frequency through DT along with the encoding of zero and one frames. Note this could be added later with another PR.

narvalotech replied:
done

simonguinot commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,246 @@
+	for (uint16_t i = 0; i < WS2812_I2S_PRE_DELAY_WORDS; i++) {
+		*tx_buf = reset_word;
+		tx_buf++;
+	}
```
Is that because the SDOUT signal is pulled high when not transmitting (it is the case on the LPC55S69), and then this could affect the first bit ?

Is there a way to configure the I2S controller to pull down the SDOUT signal when not transmitting ?

narvalotech replied:
afaik, no. at least not for the nRF series. It's the first thing I looked at. But maybe some other peripherals do, not sure if it's a spec thing or implementation thing.

simonguinot commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,246 @@
+	for (uint16_t i = 0; i < cfg->reset_words; i++) {
+		*tx_buf = reset_word;
+		tx_buf++;
+	}
```
Again if there was a way to configure the I2S controller to pull down the SDOUT signal when not transmitting, then it would become unnecessary to transmit reset words. Just waiting would be enough (as in the ws2812_spi driver).

simonguinot reviewed (changes requested):
Hi @narvalotech,

I finally managed to test your driver with an NXP LPC55S69 board and a strip with 18 Everlight B1414 LEDs. And except an issue with the reset/latch delay, I confirm it is working. Nice work :)

Please find below some additions to my previous review.

mbolivar-nordic commented:
@narvalotech sorry for the delay; it took me longer than I expected to make room for testing this in hardware.

I'm not familiar with the I2S protocol and I am having trouble getting this working. The delta to the readme related to thingy:52 and the `out-active-low` property seems to indicate a need for a pull-up resistor somewhere; where is that?

I don't see any note about that in the wiring in the blog post; the word "resistor" doesn't appear there.

For reference, I'm trying to observe the waveforms generated on SDOUT on nRF52840-DK using the following overlays:

```
───────┬────────────────────────────────────────────────────────────────────────────────
       │ File: boards/nrf52840dk_nrf52840.conf
───────┼────────────────────────────────────────────────────────────────────────────────
   1   │ CONFIG_SPI=n
   2   │ 
   3   │ CONFIG_I2S=y
   4   │ CONFIG_WS2812_STRIP=y
   5   │ CONFIG_WS2812_STRIP_I2S=y
   6   │ 
───────┴────────────────────────────────────────────────────────────────────────────────
───────┬────────────────────────────────────────────────────────────────────────────────
       │ File: boards/nrf52840dk_nrf52840.overlay
───────┼────────────────────────────────────────────────────────────────────────────────
   1   │ #include <zephyr/dt-bindings/led/led.h>
   2   │ 
   3   │ &pinctrl {
   4   │     i2s0_default_alt: i2s0_default_alt {
   5   │         group1 {
   6   │             psels = <NRF_PSEL(I2S_SCK_M, 0, 10)>,
   7   │                 <NRF_PSEL(I2S_LRCK_M, 0, 11)>,
   8   │                 <NRF_PSEL(I2S_SDOUT, 1, 13)>,
   9   │                 <NRF_PSEL(I2S_SDIN, 0, 12)>;
  10   │         };
  11   │     };
  12   │ };
  13   │ 
  14   │ i2s_led: &i2s0 {
  15   │     status = "okay";
  16   │     pinctrl-0 = <&i2s0_default_alt>;
  17   │     pinctrl-names = "default";
  18   │ };
  19   │ 
  20   │ / {
  21   │     led_strip: ws2812 {
  22   │         compatible = "worldsemi,ws2812-i2s";
  23   │ 
  24   │         i2s-dev = < &i2s_led >;
  25   │         chain-length = <16>; /* arbitrary; change at will */
  26   │         color-mapping = <LED_COLOR_ID_GREEN
  27   │                     LED_COLOR_ID_RED
  28   │                     LED_COLOR_ID_BLUE>;
  29   │         out-active-low;
  30   │         reset-delay = <120>;
  31   │     };
  32   │ 
  33   │     aliases {
  34   │         led-strip = &led_strip;
  35   │     };
  36   │ };
───────┴────────────────────────────────────────────────────────────────────────────────
```

I have the logic analyzer looking at P1.13, and my goal is to verify the timing is in spec.

Here's what I'm seeing on my console (using RTT via a separate .conf fragment not shown above):

```
[00:00:00.378,143] <inf> main: Found LED strip device ws2812
[00:00:00.378,143] <inf> main: Displaying pattern on strip
[00:00:00.379,638] <err> i2s_nrfx: Cannot write in state: 3
[00:00:00.379,669] <err> ws2812_i2s: Failed to write data: -5
[00:00:00.379,699] <err> main: couldn't update strip: -5
```

Looking briefly at the I2S headers, state 3 is I2S_STATE_STOPPING. Just including that in case it helps.

Thanks for any advice.

narvalotech commented on drivers/led_strip/ws2812_i2s.c:
```diff
@@ -0,0 +1,246 @@
+
+#define WS2812_RESET_DELAY_US(idx)    DT_INST_PROP(idx, reset_delay)
+/* Rounds up to the next 20us. */
+#define WS2812_RESET_DELAY_WORDS(idx) ((WS2812_RESET_DELAY_US(idx) + (20 - 1)) / 20)
+
+#define WS2812_NUM_COLORS(idx) (DT_INST_PROP_LEN(idx, color_mapping))
+
+#define WS2812_I2S_NUM_PIXELS(idx) (DT_INST_PROP(idx, chain_length))
+
+#define WS2812_I2S_BUFSIZE(idx)                                                                    \
+	((WS2812_NUM_COLORS(idx) * WS2812_I2S_NUM_PIXELS(idx) + WS2812_I2S_PRE_DELAY_WORDS +       \
+	  WS2812_RESET_DELAY_WORDS(idx)) *                                                         \
+	 4)
```
Yeah, my math was off by a factor of two. But this x4 multiplier is correct. With all the `word` terminology, I confused myself when calculating that reset value. I've updated the driver to only refer to `word` as a 32-bit integer.

narvalotech commented:
@mbolivar-nordic Thanks for looking at it and sorry for the big delay, I had a bunch of stuff on my bench to finish first (same as you I guess haha).

The resistor is between 5V and M1.D if you are using the thingy overlay. You also need to short M1.S to GND. This is only needed _if you don't have a logic level translator IC_. 
Basically I figured out that the thingy has this mosfet with its drain and source routed out to a connector; so if I connect it this way and invert the signal polarity (hence the `out-active-low` prop), I have myself a poor-man's level shifter.

I added a wiring comment to the overlay.

I think I saw this error from the I2S driver yeah. Do you at least get the very first LED to light up?

narvalotech commented:
@mbolivar-nordic alright figured it out.
reproduced by removing the delay in `main.c`.
Two problems:
- my wait time calculation was way off, fixed it
- the driver/peripheral seems to need some extra time for bookkeeping, so I added an `extra-wait-time` dt prop for that.

mbolivar-nordic commented:
> @mbolivar-nordic Thanks for looking at it and sorry for the big delay, I had a bunch of stuff on my bench to finish first (same as you I guess haha).

No worries! I knew that me finally putting away all the equipment to test your PR this morning would make you reappear and respond to my questions ;) (this is actually true...).

I didn't get anything to light up, no, and nothing happened on my logic analyzer. I'm not sure why a level shifter is needed, though. At least on the WS2812 strips I have, as long as the supply rail is compatible with a high pulse level, there's no problem with decoding the signal -- the lights are simply dimmer. Are you saying I should be able to remove `out-active-low` an expect to see the non-inverted signal with your latest push?

narvalotech commented:
Haha sometimes it do be like dat :p
Yeah uncommenting th out-active-low should do the trick.

Supplying the ws leds at <5V is out of spec, the newer versions might handle it better than my old strips though. So I always have 5V power and data, hence the level shifter.

mbolivar-nordic commented:
Still no luck even without the `out-active-low;` line. I get:

```
*** Booting Zephyr OS build zephyr-v3.1.0-4287-g7d55a6b8be96  ***
[00:00:00.385,559] <dbg> ws2812_i2s: ws2812_i2s_init: Word clock: freq 100000 Hz period 10 us
[00:00:00.385,589] <inf> i2s_nrfx: I2S MCK frequency: 3200000, actual PCM rate: 100000
[00:00:00.391,357] <inf> main: Found LED strip device ws2812
[00:00:00.391,387] <inf> main: Displaying pattern on strip
[00:00:00.442,718] <err> i2s_nrfx: Cannot write in state: 3
[00:00:00.442,749] <err> ws2812_i2s: Failed to write data: -5
[00:00:00.442,779] <err> main: couldn't update strip: -5
```

and the errors loop.

mbolivar-nordic commented:
@narvalotech here's an idea for how we can move this forward: can you write a boards/<BOARD>.conf and boards/<BOARD>.overlay for an official nordic DK board, and get it tested on your bench? I think I have basically all the DKs, so I should be able to reproduce on my logic analyzer here at home.

narvalotech commented:
Just Works (TM)
nrf52840dk_nrf52840.overlay
```
#include <zephyr/dt-bindings/led/led.h>

&pinctrl {
	i2s0_default_alt: i2s0_default_alt {
		group1 {
			psels = <NRF_PSEL(I2S_SCK_M, 1, 4)>,
				<NRF_PSEL(I2S_LRCK_M, 1, 5)>,
				<NRF_PSEL(I2S_SDOUT, 1, 3)>,
				<NRF_PSEL(I2S_SDIN, 1, 6)>;
		};
	};
};

i2s_led: &i2s0 {
	status = "okay";
	pinctrl-0 = <&i2s0_default_alt>;
	pinctrl-names = "default";
};

/ {
	led_strip: ws2812 {
		compatible = "worldsemi,ws2812-i2s";

		i2s-dev = < &i2s_led >;
		chain-length = <11>; /* arbitrary; change at will */
		color-mapping = <LED_COLOR_ID_GREEN
					LED_COLOR_ID_RED
					LED_COLOR_ID_BLUE>;
		reset-delay = <120>;
	};

	aliases {
		led-strip = &led_strip;
	};
};
```

nrf52840dk_nrf52840.conf
```
CONFIG_SPI=n

CONFIG_I2S=y
CONFIG_WS2812_STRIP=y
CONFIG_WS2812_STRIP_I2S=y
```

simonguinot reviewed (commented):
Hi @narvalotech,

I confirm that with your changes the sample and driver are now working out of the box on a NXP LPC55S69 board with a strip of 18 Everlight B1414 LEDs connected to the I2S1 (flexcomm7) port.

I only have a couple of minor comments. Please see them below.

simonguinot commented:
Hi @mbolivar-nordic. Did you succeed to make it work ?

simonguinot reviewed (approved):
Nice work !

mbolivar-nordic commented:
Hi all, I've definitely got this in my queue and I haven't forgotten it!

My upstream priorities have been focused on the upcoming 3.3 release,
however. I will return to this PR when the merge window opens up again
next week, and assuming I can get it working on my desk, we should be
able to get it merged promptly.

Thanks for your patience!
